src/uct/ib/rdmacm/rdmacm_iface.c

```diff
@@ -70,6 +71,230 @@ static uct_iface_ops_t uct_rdmacm_iface_ops = {
+        conn_param.private_data     = iface->ep->priv_data;
+        conn_param.private_data_len = sizeof(uct_rdmacm_priv_data_hdr_t) +
+                                      hdr->length;
+
+        if (rdma_connect(event->id, &conn_param)) {
+            ucs_error("rdma_connect(to addr=%s) failed: %m",
+                      ucs_sockaddr_str(remote_addr, ip_str, ip_len));
+            uct_rdmacm_client_handle_failure(iface, UCS_ERR_IO_ERROR);
+        }
+        break;
+
+    case RDMA_CM_EVENT_CONNECT_REQUEST:
+        /* Server - handle a connection request from the client */
+        ucs_assert(iface->is_server);
+        uct_rdmacm_iface_process_conn_req(iface, event, remote_addr);
+        break;
+
+    case RDMA_CM_EVENT_REJECTED:
+        /* Client - server rejected the connection request */
+        ucs_error("Connection request to %s rejected",
+                  ucs_sockaddr_str(remote_addr, ip_str, ip_len));
+
+        uct_rdmacm_client_handle_failure(iface, UCS_ERR_IO_ERROR);
+        break;
+
+    case RDMA_CM_EVENT_ESTABLISHED:
+        /* Client - connection is ready, call the callback to notify the upper layer */
+        ucs_assert(iface->ep != NULL);
+        hdr = (uct_rdmacm_priv_data_hdr_t*)event->param.ud.private_data;
+        /* TODO check the ep's cb_flags to determine when to invoke this callback.
+         * currently only UCT_CB_FLAG_ASYNC is supported so the cb is invoked from here */
+        iface->ep->conn_reply_cb(iface->ep->conn_reply_arg,
+                                 event->param.ud.private_data + sizeof(uct_rdmacm_priv_data_hdr_t),
+                                 hdr->length);
+
+        uct_rdmacm_iface_client_start_next_ep(iface);
+        break;
+
+    /* client error events */
+    case RDMA_CM_EVENT_ADDR_ERROR:
+    case RDMA_CM_EVENT_ROUTE_ERROR:
+    case RDMA_CM_EVENT_CONNECT_RESPONSE:
+    case RDMA_CM_EVENT_UNREACHABLE:
+
+    /* client and server error events */
+    case RDMA_CM_EVENT_CONNECT_ERROR:
+    case RDMA_CM_EVENT_DISCONNECTED:
+        /* Server/Client - connection was disconnected */
+        ucs_error("Received event %s. status = %d. Peer: %s.",
+                  rdma_event_str(event->event), event->status,
+                  ucs_sockaddr_str(remote_addr, ip_str, ip_len));
+
+        if (!iface->is_server) {
+            uct_rdmacm_client_handle_failure(iface, UCS_ERR_IO_ERROR);
+        }
+        break;
+
+    default:
+        ucs_warn("Unexpected RDMACM event: %d", event->event);
+        break;
+    }
+}
+
+static void uct_rdmacm_iface_event_handler(int fd, void *arg)
+{
+    uct_rdmacm_iface_t *iface = arg;
+    struct rdma_cm_event *event;
+    int ret;
+
+    for (;;) {
+        /* Fetch an event */
+        ret = rdma_get_cm_event(iface->event_ch, &event);
+        if (ret) {
+            /* EAGAIN (in a non-blocking rdma_get_cm_event) means that
+             * there are no more events */
+            if (errno != EAGAIN) {
```

Comment:
should also handle EINTR same way as EAGAIN

Reply:
isn't EINTR relevant only for blocking functions?